### PR TITLE
CHG: Demeaning after tapering (only) for Granger and additional front…

### LINKED ISSUE #194: Improve Granger interface
After numerical explorations of the CSD regularity and our nice discussion with Ding we should:
- check for nChan / nTrial ratio << 1 
- tighten the demeaning, which has to be done again after tapering to get rid of the *noses* at 0Hz

### PR DESCRIPTION
**Pull request by tensionhead**
…end checks

- Granger-Geweke needs full frequency range [0, Nyquist], this gets now
enforced in the connectivity fronted
- user gets warned if nChannels / nTrials is not small
- optional demeaning after tapering for mtmfft, only needed for
Granger so far, actually gives problems for coherence estimates at the 0-frequency
- resolves #194

On branch granger_improvements
Changes to be committed:
	modified:   syncopy/nwanalysis/ST_compRoutines.py
	modified:   syncopy/nwanalysis/connectivity_analysis.py
	modified:   syncopy/nwanalysis/csd.py
	modified:   syncopy/specest/mtmfft.py
	modified:   syncopy/tests/local_spy.py
	modified:   syncopy/tests/test_connectivity.py

Author Guidelines
-----------------
- [x] Is the change set **< 400 lines**?
These changes should not effect any of this
- [ ] ~Was the code checked for memory leaks/performance bottlenecks?~
- [ ] ~Is the code running locally **and** on the ESI cluster?~
- [ ] ~Is the code running on all supported platforms?~

Reviewer Checklist
------------------
- [ ] Are testing routines present?
- [ ] Do **parallel** loops have a set length and correct termination conditions?
- [ ] Do objects in the global package namespace perform proper parsing of their input? 
- [ ] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [ ] Code layout
  - [ ] Is the code PEP8 compliant?
  - [ ] Does the code adhere to agreed-upon naming conventions?
  - [ ] Are keywords clearly named and easy to understand?
  - [ ] No commented-out code?
- [ ] Are all docstrings complete and accurate?

### CODE REVIEW

File: syncopy/nwanalysis/connectivity_analysis.py [hunk 1]
```diff
@@ -240,11 +253,18 @@ def connectivityanalysis(data, method="coh", keeptrials=False, output="abs",
...
+            demean_taper = True
+        else:
+            demean_taper = False
+
         # parallel computation over trials
         st_compRoutine = ST_CrossSpectra(samplerate=data.samplerate,
                                          nSamples=nSamples,
                                          taper=taper,
                                          taper_opt=taper_opt,
+                                         demean_taper=demean_taper,
```
Comment (pantaray) [inline syncopy/nwanalysis/connectivity_analysis.py:267]: Nit-picky comment: we could shorten this to 
```
st_compRoutine = ST_CrossSpectra(samplerate=data.samplerate, ..., demean_taper=method=='granger', ...)
```

File: syncopy/specest/mtmfft.py [hunk 2]
```diff
@@ -97,9 +98,13 @@ def mtmfft(data_arr, samplerate, nSamples=None, taper="hann", taper_opt=None):
...
         win = np.tile(win, (nChannels, 1)).T
+        tapered = data_arr * win
+        # de-mean again after tapering - needed for Granger!
+        if demean_taper:
+            tapered = tapered - tapered.mean(axis=0)
         # real fft takes only 'half the energy'/positive frequencies,
         # multiply by 2 to correct for this
-        ftr[taperIdx] = 2 * np.fft.rfft(data_arr * win, n=nSamples, axis=0)
+        ftr[taperIdx] = 2 * np.fft.rfft(tapered, n=nSamples, axis=0)
         # normalization
```
Comment (pantaray) [inline syncopy/specest/mtmfft.py:108]: For memory-efficiency I'd suggest refactoring this a little:
```
win = np.tile(win, (nChannels, 1)).T
win *= data_arr
if demean_taper:
    win -= win.mean(axis=0)
ftr[taperIdx] = 2 * np.fft.rfft(win, n=nSamples, axis=0)
```

**Review by pantaray — CHANGES_REQUESTED** [review]
Nice work @tensionhead I'd just modify the `data_arr` gymnastics in `mtmfft` a little to not allocate an additional `nChan x nSample` array in there

File: syncopy/nwanalysis/connectivity_analysis.py (hunk 1 above)
Comment (tensionhead) [reply syncopy/nwanalysis/connectivity_analysis.py:267]: uff, never seen that before it would work of course..

File: syncopy/specest/mtmfft.py (hunk 2 above)
Comment (tensionhead) [reply syncopy/specest/mtmfft.py:108]: Done it latest commit!

**Review by pantaray — APPROVED** [review]
Looks great!
